unix_integration/pam_kanidm/src/pam/mod.rs

```diff
@@ -188,94 +188,136 @@ impl PamHooks for PamKanidm {
             }
         };
 
-        let authtok = match pamh.get_authtok() {
-            Ok(atok) => atok,
+        let cfg = match get_cfg() {
+            Ok(cfg) => cfg,
+            Err(e) => return e,
+        };
+        let mut req = ClientRequest::PamAuthenticateInit(account_id);
+        let mut prompt: PamPrompt = Default::default();
+
+        let mut authtok = match pamh.get_authtok() {
+            Ok(Some(v)) => Some(v),
+            Ok(None) => {
+                if opts.use_first_pass {
+                    if opts.debug {
+                        println!("Don't have an authtok, returning PAM_AUTH_ERR");
+                    }
+                    return PamResultCode::PAM_AUTH_ERR;
+                }
+                None
+            }
             Err(e) => {
                 if opts.debug {
                     println!("Error get_authtok -> {:?}", e);
                 }
                 return e;
             }
         };
+        let conv = match pamh.get_item::<PamConv>() {
+            Ok(conv) => conv,
+            Err(err) => {
+                if opts.debug {
+                    println!("Couldn't get pam_conv");
+                }
+                return err;
+            }
+        };
 
-        let authtok = match authtok {
-            Some(v) => v,
-            None => {
-                if opts.use_first_pass {
-                    if opts.debug {
-                        println!("Don't have an authtok, returning PAM_AUTH_ERR");
+        loop {
+            let timeout = match prompt.timeout {
+                Some(timeout) => timeout,
+                None => cfg.unix_sock_timeout,
+            };
+            match call_daemon_blocking(cfg.sock_path.as_str(), &req, timeout) {
```

Comment:
I just discovered in testing that `call_daemon_blocking` disconnects and starts a new session with the daemon each time it's called, destroying our pam state data. This causes us to fail calling `PamAuthenticateStep` because the the state is uninitialized again.